File: Python/106.Construct Binary Tree from Inorder and Postorder Traversal.py
#根据二叉树的后序遍历和中序遍历复原二叉树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        return self.buildSubTree(postorder,inorder,0,len(postorder)-1,0,len(inorder)-1)

    def buildSubTree(self,postorder,inorder,l,r,ll,rr):
        if l>r or ll>rr:
            return None
        node = TreeNode(postorder[r])
        i = ll
        while i<=rr:
            if inorder[i]==postorder[r]:
                break
            i += 1
        length1 = i-ll-1    ##表示分为左子树的长度
        length2 = rr-i-1    ##表示分为右子树的长度
        node.left = self.buildSubTree(postorder,inorder,l,length1+l,ll,i-1)
        node.right = self.buildSubTree(postorder,inorder,r-length2-1,r-1,i+1,rr)
        return node

class Ssolution(object):
    def inorderTraversal(self, root):
        self.result = []
        self.inorder(root)
        return self.result

    def inorder(self,p):
        if p!=None:
            self.inorder(p.left) if p.left!=None else None
            self.result.append(p.val)
            self.inorder(p.right) if p.right!=None else None

ss = Ssolution()
s = Solution()
root = s.buildTree([4,2,8,9,6,1,7,3,5],[4,9,8,6,2,7,5,3,1])
print(ss.inorderTraversal(root))

File: C++/Tree/106.Construct Binary Tree from Inorder and Postorder Traversal.cpp
/*
给出二叉树的后序和中序遍历，求出二叉树。
中序：左子树，根节点，右子树
后序：左子树，右子树，根节点
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<int> mypostorder;
	vector<int> myinorder;
public:
	TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.size() == 0)	return NULL;
		mypostorder = postorder, myinorder = inorder;
		return buildSubTree(0, postorder.size() - 1, 0, inorder.size() - 1);
    }
    TreeNode* buildSubTree(int postleft,int postright,int inleft,int inright)
    {
    	if(postleft>postright || inleft>inright)	return NULL;
    	TreeNode *root = new TreeNode(mypostorder[postright]);
    	for(int i=0; i <= inright; ++i)
    	{
    		if(myinorder[i]==root->val)
    		{
    			root->left = buildSubTree(postleft,postleft+i-inleft-1,inleft,i-1);
    			root->right = buildSubTree(postright-(inright-i),postright-1,i+1,inright);
    			return root;
    		}
    	}
    	return NULL;
    }
};

void preorder(TreeNode *root)
{
	if (root)
	{
		cout << root->val << endl;
		preorder(root->left);
		preorder(root->right);	
	}
}

int main()
{
	Solution s;
	vector<int> inorder{ 4,2,8,5,1,6,9,3,7 };
	vector<int> postorder{ 4,8,5,2,9,6,7,3,1 };
	TreeNode *tree = s.buildTree(inorder, postorder);
	preorder(tree);
}

